### PR TITLE
[Sprite Lab] Fix whenSpriteCreated bug

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR fixes a bug in Sprite Lab (CDO version) - the blocks in the `whenSpriteCreated` event fire regardless of the type of sprite costume they’re bound to. (Example: 'purple bunny' below.)

Screen Shot 2023-06-11 at 10 34 40 PM

Refer to the screencast video below of [this project](https://studio.code.org/projects/spritelab/dIZnEgo_G1kTiQZ1zah9tf2Tl-QXJecgbDCmJ5v68ns).

BEFORE UPDATE:

https://github.com/code-dot-org/code-dot-org/assets/107423305/98867dff-f266-4eb2-9b56-0f8378d12039

The reason for this bug is that in `runSpriteCreatedEvents` in `spritelab/CoreLibrary.js`, when both `inputEvent.args.name` and `newSprite.name` are `undefined` the condition for filtering the `inputEvents` for the `whenSpriteCreated` evaluates to `true` and the `whenSpriteCreated` event is incorrectly made. So I added a check to make sure the event name is not `undefined`. I also refactored the filtering of the input events for easier readability.

AFTER UPDATE:

https://github.com/code-dot-org/code-dot-org/assets/107423305/b3c4d2a4-03e0-4505-8386-a10fed41d90a

## Links
[jira ticket](https://codedotorg.atlassian.net/browse/SL-903)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/p5lab/spritelab/CoreLibrary.js [hunk 1]
```diff
@@ -549,13 +549,21 @@ export default class CoreLibrary {
...
-        inputEvent.type === 'whenSpriteCreated' &&
-        (inputEvent.args.name === newSprite.name ||
-          inputEvent.args.costume === newSprite.getAnimationLabel() ||
-          inputEvent.args.costume === 'all')
-    );
+    const matchingInputEvents = this.inputEvents.filter(inputEvent => {
+      if (inputEvent.type === 'whenSpriteCreated') {
+        const inputEventName = inputEvent.args.name;
+        const inputEventCostume = inputEvent.args.costume;
+        if (inputEventName !== undefined && inputEventName === newSprite.name) {
```
Comment (bencodeorg) [inline apps/src/p5lab/spritelab/CoreLibrary.js]: Feels a little weird to me that both the if and else if here return true?

Could you just replace the problematic condition in the existing code with your updated condition, or did that just feel too confusing?

**Review by bencodeorg — COMMENTED** [review]
Nice find!

File: apps/src/p5lab/spritelab/CoreLibrary.js (hunk 1 above)
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Sure. I thought the condition was getting kinda long so changed it to this way, but can just go with the longer condition which looks like this. Either way works.
```
    const matchingInputEvents = this.inputEvents.filter(
      inputEvent =>
        inputEvent.type === 'whenSpriteCreated' &&
        ((inputEvent.args.name !== undefined &&
          inputEvent.args.name === newSprite.name) ||
          inputEvent.args.costume === newSprite.getAnimationLabel() ||
          inputEvent.args.costume === 'all')
    );
```
Lmk what you think - thanks!

Comment (bencodeorg) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Yeah agree kind of long, but if we have to pick between the two I think this is clearer than the other version

**Review by bencodeorg — APPROVED** [review]

**Comment by ebeastlake** [comment]
If you wanted to increase readability, you could do something like the below, which is longer but possibly more readable. Curious about what others think! Logic is hard 😄

```
const matchingInputEvents = this.inputEvents.filter((inputEvent) => {
  // Ignore non-creation events
  if (inputEvent.type !== "whenSpriteCreated") {
    return false;
  }

  // Match events that apply to any costume or the current costume
  if (inputEvent.args.costume === "all" || inputEvent.args.costume === newSprite.getAnimationLabel()) {
    return true;
  }

  // Match events that apply to the current named sprite, if the inputEvent is name-filtered
  if (
    inputEvent.args.name !== undefined &&
    inputEvent.args.name === newSprite.name
  ) {
    return true;
  }

  return false;
});

matchingInputEvents.forEach((inputEvent) => {
  this.eventLog.push(`spriteCreated: ${newSprite.id}`);
  inputEvent.callback({ newSprite: newSprite.id });
});
```

**Review by ebeastlake — APPROVED** [review]

**Comment by fisher-alice** [comment]
> If you wanted to increase readability, you could do something like the below, which is longer but possibly more readable. Curious about what others think! Logic is hard 😄
> 
> ```
> const matchingInputEvents = this.inputEvents.filter((inputEvent) => {
>   // Ignore non-creation events
>   if (inputEvent.type !== "whenSpriteCreated") {
>     return false;
>   }
> 
>   // Match events that apply to any costume or the current costume
>   if (inputEvent.args.costume === "all" || inputEvent.args.costume === newSprite.getAnimationLabel()) {
>     return true;
>   }
> 
>   // Match events that apply to the current named sprite, if the inputEvent is name-filtered
>   if (
>     inputEvent.args.name !== undefined &&
>     inputEvent.args.name === newSprite.name
>   ) {
>     return true;
>   }
> 
>   return false;
> });
> 
> matchingInputEvents.forEach((inputEvent) => {
>   this.eventLog.push(`spriteCreated: ${newSprite.id}`);
>   inputEvent.callback({ newSprite: newSprite.id });
> });
> ```

Thanks Emily! This is very similar to my original commit [here](https://github.com/code-dot-org/code-dot-org/pull/52300/commits/0a293899e7a00a010c3befd77723eb2ffa3935da), but @bencodeorg preferred the long condition. I don't feel strongly either way.

**Pull request merged**